types/mempool/sender_nonce.go

```diff
@@ -0,0 +1,189 @@
+package mempool
+
+import (
+	crand "crypto/rand" // #nosec // crypto/rand is used for seed generation
+	"encoding/binary"
+	"fmt"
+	"math/rand" // #nosec // math/rand is used for random selection and seeded from crypto/rand
+
+	huandu "github.com/huandu/skiplist"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth/signing"
+)
+
+var (
+	_ Mempool  = (*senderNonceMempool)(nil)
+	_ Iterator = (*senderNonceMepoolIterator)(nil)
+)
+
+type senderNonceMempool struct {
+	senders map[string]*huandu.SkipList
+	rnd     *rand.Rand
+}
+
+// NewSenderNonceMempool creates a new mempool that prioritizes transactions by nonce, the lowest first.
+func NewSenderNonceMempool() Mempool {
+	senderMap := make(map[string]*huandu.SkipList)
+	snp := &senderNonceMempool{
+		senders: senderMap,
+	}
+
+	var seed int64
+	err := binary.Read(crand.Reader, binary.BigEndian, &seed)
+	if err != nil {
+		panic(err)
+	}
+	snp.setSeed(seed)
+
+	return snp
+}
+
+// NewSenderNonceMempoolWithSeed creates a new mempool that prioritizes transactions by nonce, the lowest first and sets the random seed.
+func NewSenderNonceMempoolWithSeed(seed int64) Mempool {
+	senderMap := make(map[string]*huandu.SkipList)
+	snp := &senderNonceMempool{
+		senders: senderMap,
+	}
+	snp.setSeed(seed)
+	return snp
+}
+
+func (snm *senderNonceMempool) setSeed(seed int64) {
+	s1 := rand.NewSource(seed)
+	snm.rnd = rand.New(s1) //#nosec // math/rand is seeded from crypto/rand by default
+}
+
+// Insert adds a tx to the mempool. It returns an error if the tx does not have at least one signer.
+// priority is ignored.
+func (snm *senderNonceMempool) Insert(_ sdk.Context, tx sdk.Tx) error {
+	sigs, err := tx.(signing.SigVerifiableTx).GetSignaturesV2()
+	if err != nil {
+		return err
+	}
+	if len(sigs) == 0 {
+		return fmt.Errorf("tx must have at least one signer")
+	}
+
+	sig := sigs[0]
+	sender := sig.PubKey.Address().String()
+	nonce := sig.Sequence
+	senderTxs, found := snm.senders[sender]
```

Comment:
does this mean that an account with an entry in the mempool and could be up for inclusion in the next block will have all the txs it has included?

Reply:
not necessarily

Reply:
if `PrepareProposal` reaches maxBytes they could be excluded.

Reply:
each select iteration randomly chooses a sender, while maintaining nonce order within the senders.

Reply:
thanks, should of read select instead of insert lol